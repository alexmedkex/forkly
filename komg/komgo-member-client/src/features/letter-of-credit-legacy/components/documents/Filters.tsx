import * as React from 'react'
import { Dropdown } from 'semantic-ui-react'
import styled from 'styled-components'
import { ILetterOfCredit } from '../../types/ILetterOfCredit'
import { FILTERS_NAME, Category, DocumentsFilters } from '../../../document-management/store'
import { SelectCategoryDropdown } from '../../../document-management/components/dropdowns/SelectCategoryDropdown'

interface IProps {
  filters: DocumentsFilters
  letterOfCredit: ILetterOfCredit
  categories?: Category[]
  enableSharedDropdown?: boolean
  filterDocuments(filter: string, value: string): void
}

const Filters: React.SFC<IProps> = (props: IProps) => {
  const shouldRenderParcelFilter = (): boolean => {
    const categoriesThatDoNotHaveParcel = ['trade-finance-documents', 'commercial-documents']
    return !categoriesThatDoNotHaveParcel.includes(props.filters.selectedCategoryId)
  }

  return (
    <StyledRightFilters>
      {props.categories && categoriesDropdown(props)}
      {props.enableSharedDropdown && sharedDropdown(props)}
      {shouldRenderParcelFilter() && parcelsDropdown(props)}
    </StyledRightFilters>
  )
}

const categoriesDropdown: React.SFC<IProps> = (props: IProps) => {
  return (
    <SelectCategoryDropdown
      categories={props.categories!}
      filters={props.filters}
      onCategorySelect={props.filterDocuments}
      disabled={false}
    />
  )
}

const sharedDropdown: React.SFC<IProps> = (props: IProps) => {
  const options = [
    { key: 'all', text: 'All', value: 'all' },
    { key: 'unshared', text: 'Unshared', value: 'unshared' },
    { key: 'shared', text: 'Shared', value: 'shared' }
  ]

  const handleChange = (event: React.SyntheticEvent, data: any) => {
    props.filterDocuments(FILTERS_NAME.SHARED, data.value)
  }

  return <Dropdown inline={true} options={options} defaultValue="all" button={true} onChange={handleChange} />
}

const parcelsDropdown: React.SFC<IProps> = (props: IProps) => {
  const parcels = props.letterOfCredit.tradeAndCargoSnapshot
    ? props.letterOfCredit.tradeAndCargoSnapshot.cargo.parcels
    : []

  const parcelOptions = [{ key: 'all', text: 'All Parcels', value: 'all' }]
  parcels.forEach(parcel => {
    parcelOptions.push({ key: parcel.id!, text: `Parcel ID: ${parcel.id!}`, value: parcel.id! })
  })

  const handleChange = (event: React.SyntheticEvent, data: any) => {
    props.filterDocuments(FILTERS_NAME.PARCEL, data.value)
  }
  return <Dropdown inline={true} options={parcelOptions} defaultValue="all" button={true} onChange={handleChange} />
}

const StyledRightFilters = styled.div`
  float: right;
`

export default Filters
