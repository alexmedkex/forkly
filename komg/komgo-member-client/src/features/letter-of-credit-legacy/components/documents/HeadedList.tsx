import * as React from 'react'
import { List } from 'semantic-ui-react'
import styled from 'styled-components'
import { HasId } from '../../../../features/document-management/store'
import GreyHeader from '../GreyHeader'

const FlexSpaceBetween = styled.div`
  display: flex;
  flex-direction: row;
  height: 30px;
`

interface Props<T extends HasId> {
  title: string
  items: T[]
  itemToListItemContent(item: T): { [index: string]: JSX.Element | string }
}

export function HeadedList<T extends HasId>(props: Props<T>) {
  return (
    <List divided={true}>
      <GreyHeader content={props.title} block={true} />
      {props.items.map((item, index) => {
        return (
          <List.Item
            key={`${item.id}_${index}`}
            content={pluckRenderableFields(props.itemToListItemContent(item))}
            style={{ paddingBottom: 0 }}
          />
        )
      })}
    </List>
  )
}

export function pluckRenderableFields(item: any) {
  return (
    <FlexSpaceBetween>
      {Object.keys(item).map(field => {
        return field === 'id' ? null : <React.Fragment key={`${item.id}_${field}`}>{item[field]}</React.Fragment>
      })}
    </FlexSpaceBetween>
  )
}
