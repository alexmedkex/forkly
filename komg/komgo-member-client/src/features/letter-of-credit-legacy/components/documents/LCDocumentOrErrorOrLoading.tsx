import * as React from 'react'
import { Document } from '../../../../components/document/Document'
import { Segment } from 'semantic-ui-react'
import { ErrorMessage, LoadingTransition } from '../../../../components'

export interface LCDocumentOrErrorOrLoadingProps {
  document?: string
  error?: string
}

export const LCDocumentOrErrorOrLoading: React.SFC<LCDocumentOrErrorOrLoadingProps> = (
  props: LCDocumentOrErrorOrLoadingProps
) => {
  if (props.error) {
    return <ErrorMessage error={props.error} title="Error while generating PDF document" />
  } else if (props.document) {
    return <Document base64Content={props.document} />
  } else {
    return (
      <Segment padded="very">
        <LoadingTransition title="Loading document" marginTop="0" />
      </Segment>
    )
  }
}
