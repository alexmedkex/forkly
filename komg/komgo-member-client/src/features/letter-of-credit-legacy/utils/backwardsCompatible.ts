import { ILetterOfCredit } from '../types/ILetterOfCredit'
import { ICargo, ITrade } from '@komgo/types'

export const withTradeAndCargoSnapshotSourceIds = (letterOfCredit?: ILetterOfCredit): ILetterOfCredit =>
  letterOfCredit && {
    ...letterOfCredit,
    tradeAndCargoSnapshot: letterOfCredit.tradeAndCargoSnapshot && {
      ...letterOfCredit.tradeAndCargoSnapshot,
      trade: letterOfCredit.tradeAndCargoSnapshot.trade && withSourceId(letterOfCredit.tradeAndCargoSnapshot.trade),
      cargo: letterOfCredit.tradeAndCargoSnapshot.cargo && withSourceId(letterOfCredit.tradeAndCargoSnapshot.cargo)
    }
  }

interface VaktId {
  vaktId?: string
}

const withSourceId = <T extends ITrade | ICargo>({ vaktId, ...payload }: T & VaktId): T =>
  ({
    ...payload,
    sourceId: (payload as T).sourceId || vaktId
  } as T)
