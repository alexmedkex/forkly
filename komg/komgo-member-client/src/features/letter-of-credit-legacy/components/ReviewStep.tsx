import * as React from 'react'
import { ILetterOfCredit } from '../types/ILetterOfCredit'
import { GetLetterOfCreditPayload } from '../utils/createAndGetLetterOfCredit'
import { LCDocumentOrErrorOrLoading } from './documents/LCDocumentOrErrorOrLoading'
import { AxiosResponse } from 'axios'

export interface ReviewStepProps {
  letterOfCredit: ILetterOfCredit
  getDocument: (payload: GetLetterOfCreditPayload) => Promise<AxiosResponse>
}

export interface ReviewStepState {
  document: string | undefined
  error: string | undefined
}

export class ReviewStep extends React.Component<ReviewStepProps, ReviewStepState> {
  async componentDidMount() {
    const { letterOfCredit } = this.props

    if (letterOfCredit.reference) {
      const payload: GetLetterOfCreditPayload = { id: letterOfCredit._id! }
      try {
        const { data } = await this.props.getDocument(payload)
        this.setState({
          document: Buffer.from(new Uint8Array(data)).toString('base64')
        })
      } catch (err) {
        this.setState({ error: err.message })
      }
    } else {
      this.setState({ error: 'A letter of credit with a valid reference was not provided' })
    }
  }

  public render() {
    return <LCDocumentOrErrorOrLoading {...this.state} />
  }
}

export default ReviewStep
