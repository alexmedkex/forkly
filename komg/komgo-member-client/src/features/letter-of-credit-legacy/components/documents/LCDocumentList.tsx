import * as React from 'react'
import { connect } from 'react-redux'
import { RouteComponentProps, withRouter } from 'react-router'
import { compose } from 'redux'
import styled from 'styled-components'
import { LoadingTransition } from '../../../../components'
import { ApplicationState } from '../../../../store/reducers'
import { displayDate } from '../../../../utils/date'
import {
  DOWNLOAD,
  DropdownOption,
  VIEW
} from '../../../document-management/components/documents/my-documents/DocumentListDropdownOptions'
import {
  groupBy,
  MapDocumentsToDocumentTypeId
} from '../../../document-management/components/documents/my-documents/toMap'
import { EllipsisDropdown } from '../../../document-management/components/dropdowns/EllipsisDropdown'
import { withCategories, withDocuments, withDocumentTypes, withProducts } from '../../../document-management/hoc'
import { Category, Document, DocumentsFilters, DocumentType, ProductId } from '../../../document-management/store/types'
import { initiateDownload } from '../../../document-management/utils/downloadDocument'
import { ILCPresentation } from '../../types/ILCPresentation'
import { ILetterOfCredit } from '../../types/ILetterOfCredit'
import DocumentTabPresentations from '../presentation/DocumentTabPresentations'
import { HeadedList } from './HeadedList'

interface IProps extends RouteComponentProps<{ id: string }> {
  filters: DocumentsFilters
  categories: Category[]
  documentTypes: DocumentType[]
  allDocs: Document[]
  documentsGroupByType: MapDocumentsToDocumentTypeId
  isLoadingDocuments: boolean
  isLoadingDocumentTypes: boolean
  letterOfCredit: ILetterOfCredit
  presentations: ILCPresentation[]
  company: string
  fetchDocumentsAsync(productId: ProductId, lc: string): void
  fetchCategoriesAsync(productId: ProductId): void
  fetchDocumentTypesAsync(productId: ProductId, categoryId?: string): void
  downloadDocumentsAsync(documentId: string, productId: ProductId): void
  changeDocumentsFilter(filter: string, value: string): void
}

const PRODUCT_ID = 'tradeFinance'

export type LCDocument = Document & { context: { parcelId: string } }

export class LCDocumentList extends React.Component<IProps> {
  componentDidMount() {
    this.props.fetchDocumentsAsync(PRODUCT_ID, this.props.match.params.id)
    this.props.fetchCategoriesAsync(PRODUCT_ID)
    this.props.fetchDocumentTypesAsync(PRODUCT_ID)
  }

  renderDocumentDropdownActions = (document: LCDocument): DropdownOption[] => {
    const viewClickHandler = () => {
      this.props.history.push(
        `/financial-instruments/letters-of-credit/${this.props.letterOfCredit._id}/documents/${document.id}`
      )
    }
    const download: DropdownOption = {
      ...DOWNLOAD,
      onClick: () => initiateDownload(document)
    }
    const view: DropdownOption = {
      ...VIEW,
      onClick: viewClickHandler
    }
    const dropdownOptions = { download, view }
    return Object.values(dropdownOptions)
  }

  render() {
    const { letterOfCredit, isLoadingDocuments, isLoadingDocumentTypes, presentations, allDocs, company } = this.props
    if (isLoadingDocuments || isLoadingDocumentTypes) {
      return (
        <LCDocumentListWrapper>
          <StyledLoader title="Loading documents" />
        </LCDocumentListWrapper>
      )
    }
    return (
      <LCDocumentListWrapper>
        {/* We do not need filters on this page for now, will need them later! <StyledHeader>
          <Filters
            filters={this.props.filters}
            letterOfCredit={letterOfCredit}
            filterDocuments={this.props.changeDocumentsFilter}
            categories={this.props.categories}
          />
        </StyledHeader> */}
        {this.renderLCDocumentsByCategory()}
        {company === letterOfCredit.applicantId && (
          <DocumentTabPresentations
            presentations={presentations}
            documents={allDocs}
            itemToListItemContent={this.documentToListItemContent}
            lcId={letterOfCredit._id}
          />
        )}
      </LCDocumentListWrapper>
    )
  }

  private renderLCDocumentsByCategory = () => {
    const lcDocumentsByCategory = groupBy(
      this.props.allDocs.filter(doc => !doc.context || (doc.context && !doc.context.lcPresentationStaticId)),
      doc => doc.category.id
    )
    const categoryIds = Array.from(lcDocumentsByCategory.keys()).filter(this.applyCategoryFilter)
    return categoryIds.map(categoryId => {
      const documentsForThisCategory = lcDocumentsByCategory.get(categoryId) || []
      return documentsForThisCategory.length ? (
        <HeadedList
          key={categoryId}
          title={categoryId}
          items={documentsForThisCategory.filter(this.applyDocumentFilters)}
          itemToListItemContent={this.documentToListItemContent}
        />
      ) : null
    })
  }

  private applyDocumentFilters = (document: LCDocument) => {
    return this.applyCategoryFilter(document.category.id) && this.applyParcelIdFilter(document.context.parcelId)
  }

  private applyCategoryFilter = (categoryId: string) => {
    const { filters } = this.props
    if (!filters) {
      return true
    }
    const selectedCategoryId = filters.selectedCategoryId || 'all'
    return selectedCategoryId === 'all' ? true : categoryId === selectedCategoryId
  }

  private applyParcelIdFilter = (parcelId: string) => {
    const { filters } = this.props
    if (!filters) {
      return true
    }
    const selectedParcelId = filters.parcel || 'all'
    return selectedParcelId === 'all' ? true : parcelId === selectedParcelId
  }

  private documentToListItemContent = (document: Document) => {
    return {
      id: document.id,
      documentType: <StyledDocumentType>{document.type.name}</StyledDocumentType>,
      fileName: <StyledDocumentProp>{document.name}</StyledDocumentProp>,
      parcelId: <StyledDocumentProp>{this.parcelIdOrNull(document)}</StyledDocumentProp>,
      registrationDate: <StyledDocumentProp>{displayDate(document.registrationDate)}</StyledDocumentProp>,
      controls: (
        <EllipsisDropdown
          options={this.renderDocumentDropdownActions}
          item={document}
          style={{ top: '-5px', paddingRight: '0' }}
        />
      )
    }
  }

  private parcelIdOrNull(document: Document) {
    const { context } = document
    return context && context.parcelId ? `Parcel #${context.parcelId}` : null
  }
}

const StyledHeader = styled.div`
  &:after {
    content: '';
    clear: both;
    display: table;
  }
  margin-bottom: 30px;
`

const StyledLoader = styled(LoadingTransition)`
  margin-top: 50px;
`

const LCDocumentListWrapper = styled.div`
  min-height: 40vh;
`

const StyledDocumentType = styled.b`
  flex: 1;
`

const StyledDocumentProp = styled.p`
  flex: 1;
`

const mapStateToProps = (state: ApplicationState, ownProps: IProps) => {
  const letterOfCreditId = ownProps.match.params.id

  const letterOfCredit: ILetterOfCredit =
    state
      .get('lettersOfCredit')
      .get('byId')
      .toJS()[letterOfCreditId] || {}

  let presentations: ILCPresentation[]

  if (letterOfCredit.reference) {
    presentations = state
      .get('lCPresentation')
      .get('byLetterOfCreditReference')
      .toJS()[letterOfCredit.reference]
  }

  return {
    letterOfCredit,
    presentations,
    company: state.get('uiState').get('profile').company
  }
}

export default compose<any>(
  withProducts,
  withCategories,
  withDocuments,
  withRouter,
  withDocumentTypes,
  connect(mapStateToProps)
)(LCDocumentList)
