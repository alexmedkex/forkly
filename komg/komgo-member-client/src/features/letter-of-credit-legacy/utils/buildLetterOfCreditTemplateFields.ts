import { ILetterOfCreditTemplate, LetterOfCreditValues } from '../constants'
import { IMember } from '../../members/store/types'
import { ITradeEnriched } from '../../trades/store/types'
import { replaceUnderscores } from '../../../utils/casings'
import { participantDetailsFromMember, findMembersByStatic } from './selectors'
import { displayPaymentTerms } from '../../../utils/date'
import { ICargo } from '@komgo/types'

export function buildLetterOfCreditTemplateFields(
  values: LetterOfCreditValues,
  members: IMember[],
  trade: ITradeEnriched,
  cargoMovements: ICargo[]
): ILetterOfCreditTemplate {
  const applicant = participantDetailsFromMember(findMembersByStatic(members, values.applicantId))
  const beneficiary = participantDetailsFromMember(findMembersByStatic(members, values.beneficiaryId))
  const issuingBank = participantDetailsFromMember(findMembersByStatic(members, values.issuingBankId))
  const beneficiaryBank = participantDetailsFromMember(findMembersByStatic(members, values.beneficiaryBankId))
  const paymentTerms = trade.paymentTerms
  const cargoMovement = cargoMovements[0] || { grade: undefined }
  return {
    ...values,
    issuingBankAddress: issuingBank.address,
    issuingBankCountry: issuingBank.country,
    beneficiaryBankAddress: beneficiaryBank.address,
    beneficiaryBankCountry: beneficiaryBank.country,
    applicant: {
      name: applicant.companyLegalName,
      locality: applicant.city
    },
    beneficiary: {
      name: beneficiary.companyLegalName,
      locality: beneficiary.city
    },
    beneficiaryBank: {
      name: beneficiaryBank.companyLegalName,
      locality: beneficiaryBank.city
    },
    issuingBank: {
      name: issuingBank.companyLegalName,
      locality: issuingBank.city
    },
    grade: cargoMovement.grade,
    cargo: {
      deliveryPeriod: {
        startDate: trade.deliveryPeriod.startDate,
        endDate: trade.deliveryPeriod.endDate
      },
      paymentTerms: displayPaymentTerms({
        eventBase: paymentTerms.eventBase,
        when: paymentTerms.when,
        time: paymentTerms.time,
        timeUnit: paymentTerms.timeUnit,
        dayType: paymentTerms.dayType
      }),
      price: trade.price,
      currency: trade.currency,
      priceUnit: trade.priceUnit,
      quantity: trade.quantity,
      deliveryTerms: trade.deliveryTerms,
      minTolerance: trade.minTolerance,
      maxTolerance: trade.maxTolerance
    },
    // Remove underscores
    applicableRules: replaceUnderscores(values.applicableRules),
    availableBy: replaceUnderscores(values.availableBy),
    // MM - Todo, hardcoded values provided until we have the proper values defined.
    // These did not break before, because we were not running that piece of the template,
    // but since we can now provide different grades, then this was now breaking
    LOIAllowed: true,
    LOI: values.LOI || ''
  }
}
