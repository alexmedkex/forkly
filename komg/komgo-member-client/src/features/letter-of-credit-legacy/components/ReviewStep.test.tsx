import * as React from 'react'
import { mount, shallow } from 'enzyme'
import { ReviewStep, ReviewStepProps } from './ReviewStep'
import { fakeLetterOfCredit } from '../utils/faker'
import * as renderer from 'react-test-renderer'
import { LCDocumentOrErrorOrLoading } from './documents/LCDocumentOrErrorOrLoading'

const getDocumentMock = jest.fn()
const letterOfCreditMock = fakeLetterOfCredit()
const testProps: ReviewStepProps = {
  letterOfCredit: letterOfCreditMock,
  getDocument: getDocumentMock
}

describe('ReviewStep', () => {
  it('renders correctly', () => {
    expect(renderer.create(<ReviewStep {...testProps} />).toJSON()).toMatchSnapshot()
  })

  it('calls componentDidMount', () => {
    const spy = jest.spyOn(ReviewStep.prototype, 'componentDidMount')
    mount(<ReviewStep {...testProps} />)
    expect(spy).toHaveBeenCalled()
    spy.mockReset()
    spy.mockRestore()
  })

  it('passes undefined props to LCDocumentOrErrorOrLoading by default', () => {
    const reviewStep = shallow(<ReviewStep {...testProps} />)
    expect(
      reviewStep.containsMatchingElement(<LCDocumentOrErrorOrLoading document={undefined} error={undefined} />)
    ).toBeTruthy()
  })

  it('passes document to LCDocumentOrErrorOrLoading when document is present', () => {
    const reviewStep = mount(<ReviewStep {...testProps} />)
    const spy = jest.spyOn(ReviewStep.prototype, 'componentDidMount')
    spy.mockImplementationOnce(() => null)

    reviewStep.setState({ document: 'some document' })

    expect(
      reviewStep.containsMatchingElement(<LCDocumentOrErrorOrLoading document={'some document'} error={undefined} />)
    ).toBeTruthy()
  })

  it('passes error to LCDocumentOrErrorOrLoading when error is present', () => {
    const reviewStep = mount(<ReviewStep {...testProps} />)
    const spy = jest.spyOn(ReviewStep.prototype, 'componentDidMount')
    spy.mockImplementationOnce(() => null)

    reviewStep.setState({ error: 'some error' })

    expect(
      reviewStep.containsMatchingElement(<LCDocumentOrErrorOrLoading document={undefined} error={'some error'} />)
    ).toBeTruthy()
  })

  it('sets the error state if there is no reference on the letter of credit', async () => {
    const { reference, ...letterOfCreditWithoutRefefence } = testProps.letterOfCredit
    const testPropsWithoutReference = { ...testProps, letterOfCredit: letterOfCreditWithoutRefefence }
    const reviewStep = await mount(<ReviewStep {...testPropsWithoutReference} />)

    await reviewStep.instance().componentDidMount!()

    expect(reviewStep.state('error')).toBeDefined()
  })

  it('passes id to getDocument', async () => {
    const reviewStep = await mount(<ReviewStep {...testProps} />)

    await reviewStep.instance().componentDidMount!(),
      expect(getDocumentMock).toHaveBeenCalledWith({
        id: letterOfCreditMock._id
      })
  })

  it('sets the error state if getDocument fails', async () => {
    const reviewStep = await mount(<ReviewStep {...testProps} />)
    getDocumentMock.mockRejectedValueOnce({ message: 'some error' })

    await reviewStep.instance().componentDidMount!()

    expect(reviewStep.state()).toEqual({ error: 'some error' })
  })
})
