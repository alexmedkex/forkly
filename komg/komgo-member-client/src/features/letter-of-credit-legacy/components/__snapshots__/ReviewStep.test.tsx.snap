// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewStep renders correctly 1`] = `
.c1 {
  display: inline-block;
  font-size: 14px;
  font-weight: normal;
  color: #142435;
  margin: 0;
}

.c0 {
  margin-top: 0;
  position: relative;
  background-color: rgba(255,255,255,0.85);
}

<div
  className="ui very padded segment"
>
  <div
    className="c0"
    data-test-id="loadingTransition"
  >
    <img
      className="ui mini centered image"
      src="/images/bars.svg"
    />
    <div
      className="ui header centered"
    >
      <span
        className="c1"
        color="#142435"
      >
        Loading document
      </span>
    </div>
  </div>
</div>
`;
